dali/dfu/dfurun.cpp

```diff
@@ -1199,57 +1198,65 @@ class CDFUengine: public CInterface, implements IDFUengine
                     if (!tmp.length())
                         throw MakeStringException(-1,"Source file not specified");
                     foreigncopy = false;
-                    if ((cmd==DFUcmd_copy)||multiclustermerge) {
+                    if ((cmd==DFUcmd_copy)||multiclustermerge)
+                    {
                         foreigndalinode.setown(getForeignDali(source));
                         foreigncopy = foreigndalinode.get()!=NULL;
-                        if (foreigncopy) {
+                        if (foreigncopy)
+                        {
+                            CDfsLogicalFileName lfn;
+                            lfn.set(tmp);
+                            lfn.setForeign(foreigndalinode->endpoint(),true);
+                            tmp = lfn.get();
+
                             StringBuffer fu;
                             StringBuffer fp;
-                            if (source->getForeignUser(fu,fp)) {
+                            if (source->getForeignUser(fu,fp))
+                            {
                                  foreignuserdesc.setown(createUserDescriptor());
                                  foreignuserdesc->set(fu.str(),fp.str());
                             }
                             else
                                 foreignuserdesc.set(userdesc);
                         }
                     }
-                    if (foreigncopy) {
-                        foreignfdesc.setown(queryDistributedFileDirectory().getFileDescriptor(tmp.str(),foreignuserdesc,foreigndalinode));
-                        if (!foreignfdesc) {
-                            StringBuffer s;
-                            throw MakeStringException(-1,"Source file %s could not be found in Dali %s",tmp.str(),foreigndalinode->endpoint().getUrlStr(s).str());
-                        }
-                        kind.set(foreignfdesc->queryProperties().queryProp("@kind"));
-                        oldRoxiePrefix.set(foreignfdesc->queryProperties().queryProp("@roxiePrefix"));
-                        iskey = strsame("key", kind);
-                        if (destination->getWrap()||iskey)
-                            destination->setNumPartsOverride(foreignfdesc->numParts());
-                        if (options->getPush()) {// need to set ftslave location
+                    srcFile.setown(fdir.lookup(tmp.str(),userdesc,
+                            (cmd==DFUcmd_move)||(cmd==DFUcmd_rename)||((cmd==DFUcmd_copy)&&multiclusterinsert)));
+
+                    if (!srcFile)
+                        throw MakeStringException(-1,"Source file %s could not be found",tmp.str());
+
+                    iskey = isFileKey(srcFile);
+                    if ((cmd==DFUcmd_copy) && (srcFile->querySuperFile() != nullptr) && iskey)
+                        throwError1(DFTERR_InvalidSuperindexCopy, srcFile->queryLogicalName());
```

Comment:
trivial: as tmp is same as srcFile->queryLogicalName() would be good to be consistent here and on line 1227.
And a littler further down it is set into srcName.
You could set srcName to tmp ahead of the lookup() on line 1223, and then use srcFile in all cases.

Reply:
Changed